functions/contrib/go/configmap-injector/README.md

```diff
@@ -0,0 +1,155 @@
+# configmap-injector
+
+### Overview
+
+<!--mdtogo:Short-->
+
+Inject a set of values in to a config map using a template for substitutions.
+
+<!--mdtogo-->
+
+### Synopsis
+
+<!--mdtogo:Long-->
+
+Configured using a ConfigMapInjector config. A user can specify three properties.
+
+`target` specifies a ConfigMap reference. This is where the resultant config will be added.
+`values` specifies a ConfigMap reference. This is where the setters and values for the substitution can be specified.
+`template` specifies a ConfigMap reference. This is where a user can specify the multiline strings together with the substitutions.
+
+```yaml
+apiVersion: fn.kumorilabs.io/v1alpha1
+kind: ConfigMapInjector
```

Comment:
I know this is a generalized solution where one can generate content for the target configMap from template and values. But I am curious, if you considered the approach of implementing a generator function that takes the values through a functionConfig directly and emit `ConfigMap` with the targeted content. functionConfig will become a client side abstraction (local CRD) and one can also benefit from strict checking of input values etc. ?

Do you have other use-cases as well for template and value substitution ?

Reply:
The reason why I use a config map for the input values is so that one can derive the values from a setter.  So the pipeline we use looks like this:

``` yaml
...
pipeline:
  mutators:
    - image: gcr.io/kpt-fn/apply-setters:unstable
      configMap:
        admin-email: helix-admin@swisscom.com
        domain: cnative.io
        git-cnbb-base-url: https://github.com/synax
        git-customer-base-url: https://github.com/synax
        platform: gcp
        registry: cnbb-docker-local.bin.swisscom.com
        stage: prd
    - image: synax/configmap-injector:latest
      configPath: fn-configmap-injector-argocd-cm.yaml
```

As far as I know I can't apply setters in a function config anymore, right? At least when I tried in an earlier version it didn't work anymore :). It used to work in pre 1.0.0, but since the new setter implementation is also a function, all the function configs are not part of the resource list which is processed by the apply-setter function. It is very important that we can derive the values from the apply-setter input.

Reply:
Aah...I see. The good news is that we are working on introducing `pre-processors` that will be called top-down from parent to leaf that can act on meta-resources as well. So like in this case, apply-setters can be invoked as pre-processor to update the values in the functionConfig for the generator function and when generator function will be called, it will have values substituted by `apply-setters` available in the functionConfig.

https://github.com/GoogleContainerTools/kpt/issues/2385

https://github.com/GoogleContainerTools/kpt/issues/2420 [This is being designed in the current sprint]

/cc @phanimarupaka

Reply:
@stefanhenseler It is not clear to me as to what this function trying to achieve from the example in the README.md.  Is the aim here to read the input values to apply-setters function and use them to generate other ConfigMap ? In that case, we are adding a new field called `include-meta-resources` to function declaration in mutators so that the functions will be fed with the meta resources (Kptfile and function configs). Does that make this function simpler to implement/understand ?

Reply:
@droot Ah this is awesome! I guess this would solve our immediate problem. @phanimarupaka The reason why I even started creating this function is because of one particular use case, below the example of argocd-cd. They use a configmap to configure argocd. We don't want to specify `stage`, `domain` or any other value multiple times in multiple different functionConfigs. Thats the reason why we created the values configmap, so the apply-setters function can set the input values using the setters, and then configmap injector sets the values using the template in the multiline strings.:

```yaml
apiVersion: v1
kind: ConfigMap
metadata: # kpt-merge: /argocd-cm
  name: argocd-cm
data:
  admin.enabled: "false"
  statusbadge.enabled: "false"
  url: https://argocd.helix.prd.cnative.io # kpt-set: https://argocd.helix.${stage}.${domain}
  users.anonymous.enabled: "false"
  users.session.duration: 240h
  oidc.config: |
    name: helix
    clientID: $oidc.client-id
    clientSecret: $oidc.client-secret
    issuer: https://auth.prd.cnative.io/auth/realms/my-app # kpt-set:  https://auth.${stage}.${domain}/auth/realms/my-app <= we want to be able to use setters here
    requestedScopes:
    - openid
    - profile
    - email
    - groups
  repository.credentials: |
    - passwordSecret:
        key: password
        name: git-reader
      url: https://github.com/synax #  kpt-set:  ${git-url} <= and here
      usernameSecret:
        key: username
        name: git-reader
    - passwordSecret:
        key: password
        name: git-reader
      url: https://github.com/synax #  kpt-set:  ${git-url} <= and here
      usernameSecret:
        key: username
        name: git-reader
```